File: Numerical_Methods_for_Physics_Garcia/edition2/Cpp/ge.cpp
#include "NumMeth.h"

// ge - Function to perform Gaussian elimination to solve A*x = b
//      using scaled column pivoting
// Inputs
//    A    -    Matrix A (N by N)
//    b    -    Vector b (N by 1)
// Outputs
//    x    -    Vector x (N by 1)
//  determ -    Determinant of matrix A	 (return value)
double ge(Matrix A, Matrix b, Matrix& x) {

  int N = A.nRow();	 
  assert( N == A.nCol() && N == b.nRow() && N == x.nRow() );
    
  int i, j, k;
  Matrix scale(N);	// Scale factor
  int *index;  index = new int [N+1];	// Row index list
  
  //* Set scale factor, scale(i) = max( |A(i,j)| ), for each row
  for( i=1; i<=N; i++ ) {
    index[i] = i;		   // Initialize row index list
    double scaleMax = 0.0;
    for( j=1; j<=N; j++ ) 
      scaleMax = (scaleMax > fabs(A(i,j))) ? scaleMax : fabs(A(i,j));
    scale(i) = scaleMax;
  }

  //* Loop over rows k = 1, ..., (N-1)
  int signDet = 1;
  for( k=1; k<=(N-1); k++ ) {
	//* Select pivot row from max( |A(j,k)/s(j)| )
    double ratiomax = 0.0;
	int jPivot = k;
    for( i=k; i<=N; i++ ) {
      double ratio = fabs(A(index[i],k))/scale(index[i]);
      if( ratio > ratiomax ) {
        jPivot = i;
        ratiomax = ratio;
      }
    }
	//* Perform pivoting using row index list
	int indexJ = index[k];
	if( jPivot != k ) {	          // Pivot
      indexJ = index[jPivot];
      index[jPivot] = index[k];   // Swap index jPivot and k
      index[k] = indexJ;
	  signDet *= -1;			  // Flip sign of determinant
	}
	//* Perform forward elimination
    for( i=k+1; i<=N; i++ ) {
      double coeff = A(index[i],k)/A(indexJ,k);
      for( j=k+1; j<=N; j++ )
        A(index[i],j) -= coeff*A(indexJ,j);
      A(index[i],k) = coeff;
      b(index[i]) -= A(index[i],k)*b(indexJ);
    }
  }
  //* Compute determinant as product of diagonal elements
  double determ = signDet;	   // Sign of determinant
  for( i=1; i<=N; i++ )
	determ *= A(index[i],i);

  //* Perform backsubstitution
  x(N) = b(index[N])/A(index[N],N);
  for( i=N-1; i>=1; i-- ) {
    double sum = b(index[i]);
    for( j=i+1; j<=N; j++ )
      sum -= A(index[i],j)*x(j);
    x(i) = sum/A(index[i],i);
  }
  
  delete [] index;	 // Release allocated memory
  return( determ );        
}

File: Numerical_Methods_for_Physics_Garcia/edition2/Fortran/ge.f
      real*8 function ge( AA, bb, N, Nm, x )
      integer*4 N, Nm
      real*8 AA(Nm,Nm), bb(Nm), x(Nm)
! ge - Function to perform Gaussian elimination to solve A*x = b
!      using scaled column pivoting
! Inputs
!    AA   -    Matrix A (N by N)
!    bb   -    Vector b (N by 1)
!    N    -    Dimension of A, b, and x (used)
!    Nm   -    Dimension of A, b, and x (allocated memory)
! Outputs
!    x    -    Vector x (N by 1)
!  determ -    Determinant of matrix A  (return value)

      integer*4 MAXN
      parameter( MAXN = 500 )

      integer*4 i, j, k, index(MAXN), signDet, jPivot, indexJ
      real*8 scale(MAXN), scaleMax, ratiomax, ratio, coeff
      real*8 determ, sum, A(MAXN,MAXN), b(MAXN)

      if( Nm .gt. MAXN ) then
        write(*,*) 'ERROR - Matrix is too large for ge routine'
        stop
      endif

      ! Copy matrix A and vector b so as not to modify original
      do i=1,N
        b(i) = bb(i)
        do j=1,N
          A(i,j) = AA(i,j)
        enddo
      enddo

      !* Set scale factor, scale(i) = max( |A(i,j)| ), for each row
      do i=1,N
        index(i) = i          ! Initialize row index list
        scaleMax = 0.0
        do j=1,N
          if( abs(A(i,j)) .gt. scaleMax ) then
            scaleMax = abs(A(i,j))
          endif
        enddo
        scale(i) = scaleMax
      enddo

      !* Loop over rows k = 1, ..., (N-1)
      signDet = 1
      do k=1,N-1
        !* Select pivot row from max( |A(j,k)/s(j)| )
        ratiomax = 0.0
        jPivot = k
        do i=k,N
          ratio = abs(A(index(i),k))/scale(index(i))
          if( ratio .gt. ratiomax ) then
            jPivot = i
            ratiomax = ratio
          endif
        enddo
        !* Perform pivoting using row index list
        indexJ = index(k)
        if( jPivot .ne. k ) then     ! Pivot
          indexJ = index(jPivot)
          index(jPivot) = index(k)   ! Swap index jPivot and k
          index(k) = indexJ
          signDet = -1*signDet       ! Flip sign of determinant
        endif
        !* Perform forward elimination
        do i=k+1,N
          coeff = A(index(i),k)/A(indexJ,k)
          do j=k+1,N
            A(index(i),j) = A(index(i),j) - coeff*A(indexJ,j)
          enddo
          A(index(i),k) = coeff
          b(index(i)) = b(index(i)) - A(index(i),k)*b(indexJ)
        enddo
      enddo
      !* Compute determinant as product of diagonal elements
      determ = signDet     ! Sign of determinant
      do i=1,N
        determ = determ * A(index(i),i)
      enddo

      !* Perform backsubstitution
      x(N) = b(index(N))/A(index(N),N)
      do i=N-1,1,-1
        sum = b(index(i))
        do j=i+1,N
          sum = sum - A(index(i),j)*x(j)
        enddo
        x(i) = sum/A(index(i),i)
      enddo

      ge = determ
      return
      end
